src/credentials.rs

```diff
@@ -31,55 +31,110 @@ pub struct Credentials {
     pub registry_tokens: HashMap<RegistryUri, String>,
 }
 
-impl Credentials {
-    fn location() -> eyre::Result<PathBuf> {
-        let home = home::home_dir().wrap_err("Failed to locate home directory")?;
+/// Error produced when locating the credentials file
+#[derive(Error, Debug)]
+pub enum LocateError {
+    /// The home folder where the credentials file is stored could not be determined.
+    ///
+    /// This implies that BUFFRS_HOME is not set and that the user's home directory
+    /// could not be detected.
+    #[error("BUFFRS_HOME unset and could not resolve user's home directory")]
+    MissingHome,
+}
 
-        let home = env::var("BUFFRS_HOME").map(PathBuf::from).unwrap_or(home);
+/// Error produced when checking if the credentials file exists
+#[derive(Error, Debug)]
+pub enum ExistsError {
+    /// Could not locate the credentials file
+    #[error("Failed to determine credentials location. {0}")]
+    Locate(LocateError),
+    /// Could not access the filesystem
+    #[error("IO error: {0}")]
+    Io(std::io::Error),
+}
+
+/// Error produced when loading the credentials file
+#[derive(Error, Debug)]
+pub enum ReadError {
+    /// Could not locate the credentials file
+    #[error("Failed to determine credentials location. {0}")]
+    Locate(LocateError),
+    /// Could not read the file from the filesystem
+    #[error("IO error: {0}")]
+    Io(std::io::Error),
+    /// Could not parse the TOML credentials data
+    #[error("Failed to deserialize credentials. Cause: {0}")]
+    Toml(toml::de::Error),
+}
+
+/// Error produced when updating the credentials file
+#[derive(Error, Debug)]
+pub enum WriteError {
+    /// Could not locate the credentials file
```

Comment:
these comments are totally redundant given the error itself contains the same information. The duplicative information simply increases the likelihood that it gets stale.

Reply:
I agree. But I couldn't figure out a way to turn on the doc-comments linter without it complaining about the lack of these. I wish we could generate these from the field macro or vice-versa.

Reply:
i dont think its _that_ bad to keep them. it is redundant in code, but nice for docs.rs + users in their ide

Reply:
I think this becomes moot since we'll make these types crate private. I might remove some trivial comments but I guess there's some value in duplicating the message for the sake of IDE integration for collaborators, still.

Reply:
I found this nice crate: [displaydoc](https://docs.rs/displaydoc/latest/displaydoc/). It works well with `thiserror` so I'll be converting some doc-comments to this.